### ISSUE TITLE
Refetch should not trigger a field merge

### ISSUE THREAD

**Issue by vigie** [apollographql/apollo-client#7491]
**Intended outcome:**

From the docs for refetch:

> Update the variables of this observable query, and fetch the new results

It seems wrong to me that when the new results come back it triggers the field's merge function. As we are resetting the variables the previous value for the query should be considered invalid and thrown out implicitly.

If there is not agreement on the above proposition, at the very least I would like to have a way to know that I'm in a merge function as the result of a refetch, so I can throw out the previous result explicitly.

**Actual outcome:**

The field's merge function is triggered and new results are merged with the old, creating an erroneous cache entry that results from a combination of old variable values and new.

**How to reproduce the issue:**

This is default behavior, trigger a refetch on a query whose result has a custom field merge policy. The problem is most easily identified in the context of list type fields and pagination.

**Versions**
```
  System:
    OS: macOS 10.15.7
  Binaries:
    Node: 12.18.2 - ~/.nvm/versions/node/v12.18.2/bin/node
    npm: 6.14.5 - ~/.nvm/versions/node/v12.18.2/bin/npm
  Browsers:
    Chrome: 87.0.4280.88
    Firefox: 79.0
    Safari: 14.0.2
  npmPackages:
    @apollo/client: 3.0.0-rc.4 => 3.0.0-rc.4
    apollo-angular: ^2.0.0-beta.2 => 2.0.0-beta.2
```

**Comment by adsee42** [comment]
Any update on this?

**Comment by JeffJankowski** [comment]
Also encountering this issue. I am using the new type policies and `merge` for pagination, but I also need to be able to fetch fresh results from the network and replace the old data. I agree with @vigie that a `refetch` should evict the old cache entry, or at the very least notify the `merge` function so we can handle it explicitly.

**Comment by vigie** [comment]
Btw for anyone hitting this issue I found the following acceptable workaround: use `fetchMore` instead of `refetch`. When using `fetchMore` you can pass an argument that only persists for that one call and is not written and persisted to the query like it would be with `refetch`. I pass `dropCache: true` and interpret it as necessary in the merge function.

Perhaps the maintainers want to argue that this is in fact the correct solution..? It does however call into question what the legitimate use cases for `refetch` are...

**Comment by JeffJankowski** [comment]
> Btw for anyone hitting this issue I found the following acceptable workaround: use `fetchMore` instead of `refetch`. When using `fetchMore` you can pass an argument that only persists for that one call and is not written and persisted to the query like it would be with `refetch`. I pass `dropCache: true` and interpret it as necessary in the merge function.
> 
> Perhaps the maintainers want to argue that this is in fact the correct solution..? It does however call into question what the legitimate use cases for `refetch` are...

For those of us using TypeScript and variables types generated from `apollo client:codegen`, this is a pretty painful hack. 😕

**Comment by adsee42** [comment]
My workaround is to use `fetchMore` instead of `refetch`, and check `offset` in `merge` function.
If `offset === 0`, return `incoming`, else, do custom merge.

No extra parameter is needed😋

**Comment by gaurabsarkar** [comment]
I had similar issue. Refetch would not call merge in TypePolicy and would not update cache. I have an infinite loading list and this is how I solved it.
![types](https://user-images.githubusercontent.com/16963370/104464022-79d17880-55d8-11eb-847a-7610a2c34224.png)
![query](https://user-images.githubusercontent.com/16963370/104464034-7ccc6900-55d8-11eb-9839-f5602117c57b.png)
![cache](https://user-images.githubusercontent.com/16963370/104464043-805ff000-55d8-11eb-80b7-1b45ff5a0aab.png)
![component](https://user-images.githubusercontent.com/16963370/104464046-8229b380-55d8-11eb-9b4c-4679d9f3fe29.png)

Hope this helps!

**Comment by Daavidaviid** [comment]
I have the same issue when useQuery is called multiple times and there is a merge policy, like when I display a page that has already been displayed previously in a React Native app.
The first time it works great, I can fetch more, I can refetch the page, no issue, but when go back and go to the same page again, it breaks.

Still looking for a clean way to solve this

**Comment by jtoce** [comment]
Now that I’ve looked under the hood I can tell you why. When you use `fetchMore` with the new field policies, it writes straight to the cache.

https://github.com/apollographql/apollo-client/blob/4fd6824607b06d38e62753035fd0b3ebf1d359e1/src/core/ObservableQuery.ts#L316-L325

_However_... When you call `refetch` it instead performs the fetch (usually defaulting back to `fetchPolicy: 'network-only'`) and then it does a comparison of your previous `lastWrite` to see if anything has changed before writing to the cache.

https://github.com/apollographql/apollo-client/blob/4fd6824607b06d38e62753035fd0b3ebf1d359e1/src/core/QueryInfo.ts#L327-L364

Here's the problem- `lastWrite` is only updated on the initial fetch, but `diff` is updated by `fetchMore`. So for example- if you use pagination (i.e. page, size) and nothing has changed on your _first_ page of results, then the logic of `shouldWrite` is `false` since both your variables and data are the same. So it returns you the entire cache `diff` instead of your new network results. The obvious problem of comparing the initial fetch (i.e. page one) and then deciding to return all pages in the cache, is that something might have changed on later pages.

**Comment by benjamn** [comment]
@jtoce Thanks for digging into that! I think this should be fixable, given your analysis.

**Comment by benjamn** [comment]
Here's my current thinking on the original issue, as reported by @vigie and echoed by @JeffJankowski:

If you define a `merge` function for particular field, `InMemoryCache` must call that function any time new data is written for that field, even the very first time (when the `existing` data is undefined), because `merge` functions give you the ability to customize the internal representation of field data within the cache. If `merge` was called most of the time but not all the time, it would be very difficult to know what type of data to expect when reading from the cache.

However, I believe we can preserve this guarantee (that `merge` is always called) while achieving the behavior @vigie is proposing, by providing a way to call `merge` functions with undefined `existing` data, so they behave exactly as they would if there was no existing data, and thus do not attempt to combine incoming data with existing data (but still potentially transform the `incoming` data, as appropriate).

My current plan is to add an `overwrite?: boolean` option to the `cache.writeQuery` and `cache.writeFragment` option types. It will be `false` by default, but passing `overwrite: true` will prevent any `merge` functions involved in the write from seeing any existing data, thereby overwriting those field values. For fields that do not have a `merge` function, new field data always replaces existing data, so `overwrite: true` is a way to achieve that default behavior even when you have a `merge` function.

Of course, since you don't manually call `cache.writeQuery` when you're using `refetch`, we will need to arrange for `overwrite: true` to be passed behind the scenes, when `refetch` does its cache write. I agree this should be the default behavior for `refetch`, but we may need to make it optional (opt-in) to avoid breaking changes.

I welcome your thoughts/questions about this idea. It is straightforward to implement, but needs more unit tests. I'll kick off a PR tomorrow, with the goal of releasing it in [Apollo Client 3.4](https://github.com/apollographql/apollo-client/pull/7399).

The issue @jtoce identified in https://github.com/apollographql/apollo-client/issues/7491#issuecomment-767985363 may well be a contributing factor (preventing writes in some cases, thereby not triggering any `merge` functions), but that should be easy to fix by deleting `queryInfo.lastWrite` before refetching. I'll put that change in the PR too.

**Comment by benjamn** [comment]
@vigie @JeffJankowski PR in progress—sorry for the wait! #7810

**Comment by Akryum** [comment]
Really need this to be fixed, as I can't both use `fetchMore` and `refetch` while migrating from Apollo Client 2 to 3, `refetch` being broken (the `merge` logic don't have the info if the cache write is a pagination or a refetch). :sweat_smile: :pray:

**Comment by benjamn** [comment]
This should be fixed in the v3.4 betas, thanks to #7966 (and thanks to feedback from @Akryum).
